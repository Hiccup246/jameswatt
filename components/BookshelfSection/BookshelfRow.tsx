import React from "react";
import Image from "next/image";
import AnimatedBook from "../../public/animated-book.gif";
import ReadBook from "../../public/book.webp";

export default function BookshelfRow({ book }: { book: Book }) {
  const completedBookIcon = (
    <div className="h-5 w-5 mx-auto">
      <Image src={ReadBook}
             width={20}
             height={20}
             sizes="20px"
            alt="Completed Book Icon" />
    </div>
  )

  const currentlyReadingIcon = (
    <div className="h-5 w-5 mx-auto">
      <Image src={AnimatedBook}
              width={20}
              height={20}
              sizes="20px"
              alt="Currently Reading Icon" />
    </div>
  )

  return (
    <>
      <td className="flex justify-start flex-col pr-2.5 break-keep align-top">
        <h2 className="my-0 max-w-lg font-bold">{book.name}</h2>
        <div className="text-sm">{book.author}</div>
      </td>

      <td className="align-top text-sm max-smaller:inline">{book.genre}</td>
      
      <td className="align-top">
        {
          book.status == "Read" ? completedBookIcon : currentlyReadingIcon
        }
      </td>
    </>
  )
}