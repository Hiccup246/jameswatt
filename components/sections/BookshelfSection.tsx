import { useState } from "react";
import { BOOKS } from "../../constants/Books";
import BookShelfTable from "../BookshelfTable";
import SectionLayout from "../layouts/SectionLayout";

export function bookStatusSorter(a: Book, b: Book): number {
  if (a.status == "Reading" && b.status != "Reading") return -1;
  if (a.status != "Reading" && b.status == "Reading") return 1;

  return 0;
}

const sortedBooks = BOOKS.slice().sort(bookStatusSorter);

export default function BookshelfSectionTwo() {
  const [showAllBooks, setshowAllBooks] = useState(false);

  return (
    <SectionLayout bgSecondary>
      <h1 className="mb-16 text-center text-2xl font-bold sm:text-3xl">
        Books I&apos;m Involved With
      </h1>

      <div className="rounded-xl p-3.5">
        <div className="max-h-[600px] overflow-y-scroll pr-2">
          <BookShelfTable
            books={sortedBooks}
            shortViewBooks={4}
            shortView={!showAllBooks}
          />
        </div>

        <button
          aria-label={showAllBooks ? "Show Less Books" : "Show More Books"}
          className="mx-auto mt-8 block w-fit rounded-md border px-2 py-1 hover:opacity-50"
          onClick={() => setshowAllBooks(!showAllBooks)}
        >
          {showAllBooks ? "Show Less" : "Show More"}
        </button>
      </div>
    </SectionLayout>
  );
}
