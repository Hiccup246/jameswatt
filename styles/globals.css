@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
  :root {
    --slider-items: 7;
    --slider-item-sets: 2;
    --all-slider-items: calc(var(--slider-items) * var(--slider-item-sets));
    --slider-item-width: 200px;
  }

  html,
  body {
    padding: 0;
    margin: 0;
    font-family: ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont,
      "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif,
      "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol",
      "Noto Color Emoji";
    @apply bg-brown;
  }

  @property --gradient-angle {
    syntax: "<angle>";
    inherits: false;
    initial-value: 1turn;
  }
}

@layer utilities {
  .animation-pause {
    animation-play-state: paused;
  }

  .animation-play {
    animation-play-state: running;
  }
}
