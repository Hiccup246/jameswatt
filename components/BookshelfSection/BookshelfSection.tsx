import React, { useState } from 'react'

import { BOOKS } from '../../constants/Books'
import SectionLayout from "../SectionLayout";

// import './bookshelf.css'
import BookshelfRow from './BookshelfRow';
export default function BookshelfSection() {
    const previewBooks = 4

    const renderBooks = (numberOfBooks: number) => {
        return BOOKS.slice(0, numberOfBooks).map((book: Book, index) => {
            return (
                <tr key={index} className="[&>td]:pt-10 [&>td]:first:pt-0">
                    <BookshelfRow book={book} />
                </tr>
            )
        })
    }

    const [buttonText, setButtonText] = useState('Show More')
    const [books, setBooks] = useState(renderBooks(previewBooks))

    const showHideBooks = () => {
        if (buttonText === 'Show More') {
            setBooks(renderBooks(BOOKS.length))
            setButtonText('Show Less')
        } else {
            setBooks(renderBooks(previewBooks))
            setButtonText('Show More')
        }
    }

    return (
        <SectionLayout isBgColorPrimary={true}>
            <h1 className="self-center text-2xl sm:text-3xl font-bold mb-14">
                Books I'm Involved With
            </h1>

            <div className="bookshelf__table-wrapper bg-white p-3.5 rounded-xl">
                <table className="bookshelf__table w-full">
                    <thead>
                        <tr>
                            <th className="text-sm pb-2.5 text-left text-black">Book</th>
                            <th className="bookshelf__genre-header text-sm pb-2.5 text-left text-black w-24 max-smaller:hidden">Genre</th>
                            <th className="bookshelf__status-header text-sm pb-2.5 text-center text-black">Status</th>
                        </tr>
                    </thead>
                    <tbody>{books}</tbody>
                </table>

                <button
                    aria-label="Show More Books"
                    className="bookshelf__show-more-button border rounded-md px-2 py-1 mx-auto w-fit block hover:opacity-50 mt-8"
                    onClick={() => showHideBooks()}
                    >
                    {buttonText}
                </button>
            </div>
        </SectionLayout>
    )
};
