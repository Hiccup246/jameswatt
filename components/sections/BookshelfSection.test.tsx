import { expect, it } from "@jest/globals";
import BookshelfSection from "./BookshelfSection";
import renderer from "react-test-renderer";

it("renders the BookshelfSection correctly", () => {
  const tree = renderer.create(<BookshelfSection />).toJSON();
  expect(tree).toMatchSnapshot();
});
