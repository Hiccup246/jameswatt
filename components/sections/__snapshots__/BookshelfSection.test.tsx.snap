// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookshelfSection renders correctly 1`] = `
<div>
  <div
    class="flex w-full flex-row justify-center pt-12 pb-20 pt-20 bg-brown dark:bg-darkgrey text-black dark:text-softwhite"
  >
    <section
      class="mx-auto flex w-9/12 flex-col justify-start"
    >
      <h1
        class="mb-16 text-center text-2xl font-bold sm:text-3xl"
      >
        Books I'm Involved With
      </h1>
      <div
        class="rounded-xl p-3.5"
      >
        <div
          class="max-h-[600px] overflow-y-scroll pr-2"
        >
          <table
            class="w-full"
          >
            <thead
              class="sticky top-0 bg-brown dark:bg-darkgrey"
            >
              <tr>
                <th
                  class="pb-2.5 text-left text-sm"
                >
                  Book
                </th>
                <th
                  class="w-24 pb-2.5 text-left text-sm max-smaller:hidden"
                >
                  Genre
                </th>
                <th
                  class="pb-2.5 text-center text-sm"
                >
                  Status
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="[&>td]:pt-10 [&>td]:first:pt-0"
              >
                <td
                  class="flex flex-col justify-start break-keep pr-2.5 align-top"
                >
                  <h2
                    class="my-0 max-w-lg font-bold"
                  >
                    Whispers Under Ground
                  </h2>
                  <div
                    class="flex flex-col"
                  >
                    <div
                      class="text-sm"
                    >
                      Ben Aaronovitch
                    </div>
                    <div
                      class="order-last text-sm max-smaller:order-first"
                      title="Date Started"
                    >
                      16/02/2025
                    </div>
                  </div>
                </td>
                <td
                  class="align-top text-sm max-smaller:inline"
                >
                  Urban Fantasy
                </td>
                <td
                  class="align-top"
                >
                  <div
                    class="mx-auto h-5 w-5"
                  >
                    <img
                      alt="Currently Reading Icon"
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/animated-book-black.gif"
                      style="color: transparent;"
                      width="20"
                    />
                  </div>
                </td>
              </tr>
              <tr
                class="[&>td]:pt-10 [&>td]:first:pt-0"
              >
                <td
                  class="flex flex-col justify-start break-keep pr-2.5 align-top"
                >
                  <h2
                    class="my-0 max-w-lg font-bold"
                  >
                    Moon Over Soho
                  </h2>
                  <div
                    class="flex flex-col"
                  >
                    <div
                      class="text-sm"
                    >
                      Ben Aaronovitch
                    </div>
                    <div
                      class="order-last text-sm max-smaller:order-first"
                      title="Date Completed"
                    >
                      15/02/2025
                    </div>
                  </div>
                </td>
                <td
                  class="align-top text-sm max-smaller:inline"
                >
                  Urban Fantasy
                </td>
                <td
                  class="align-top"
                >
                  <div
                    class="mx-auto h-5 w-5"
                  >
                    <img
                      alt="Completed Book Icon"
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      sizes="20px"
                      src="/_next/image?url=%2Fclipboard-checked-black.webp&w=3840&q=75"
                      srcset="/_next/image?url=%2Fclipboard-checked-black.webp&w=16&q=75 16w, /_next/image?url=%2Fclipboard-checked-black.webp&w=32&q=75 32w, /_next/image?url=%2Fclipboard-checked-black.webp&w=48&q=75 48w, /_next/image?url=%2Fclipboard-checked-black.webp&w=64&q=75 64w, /_next/image?url=%2Fclipboard-checked-black.webp&w=96&q=75 96w, /_next/image?url=%2Fclipboard-checked-black.webp&w=128&q=75 128w, /_next/image?url=%2Fclipboard-checked-black.webp&w=256&q=75 256w, /_next/image?url=%2Fclipboard-checked-black.webp&w=384&q=75 384w, /_next/image?url=%2Fclipboard-checked-black.webp&w=640&q=75 640w, /_next/image?url=%2Fclipboard-checked-black.webp&w=750&q=75 750w, /_next/image?url=%2Fclipboard-checked-black.webp&w=828&q=75 828w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1080&q=75 1080w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1200&q=75 1200w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1920&q=75 1920w, /_next/image?url=%2Fclipboard-checked-black.webp&w=2048&q=75 2048w, /_next/image?url=%2Fclipboard-checked-black.webp&w=3840&q=75 3840w"
                      style="color: transparent;"
                      width="20"
                    />
                  </div>
                </td>
              </tr>
              <tr
                class="[&>td]:pt-10 [&>td]:first:pt-0"
              >
                <td
                  class="flex flex-col justify-start break-keep pr-2.5 align-top"
                >
                  <h2
                    class="my-0 max-w-lg font-bold"
                  >
                    Rivers of London
                  </h2>
                  <div
                    class="flex flex-col"
                  >
                    <div
                      class="text-sm"
                    >
                      Ben Aaronovitch
                    </div>
                    <div
                      class="order-last text-sm max-smaller:order-first"
                      title="Date Completed"
                    >
                      11/02/2025
                    </div>
                  </div>
                </td>
                <td
                  class="align-top text-sm max-smaller:inline"
                >
                  Urban Fantasy
                </td>
                <td
                  class="align-top"
                >
                  <div
                    class="mx-auto h-5 w-5"
                  >
                    <img
                      alt="Completed Book Icon"
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      sizes="20px"
                      src="/_next/image?url=%2Fclipboard-checked-black.webp&w=3840&q=75"
                      srcset="/_next/image?url=%2Fclipboard-checked-black.webp&w=16&q=75 16w, /_next/image?url=%2Fclipboard-checked-black.webp&w=32&q=75 32w, /_next/image?url=%2Fclipboard-checked-black.webp&w=48&q=75 48w, /_next/image?url=%2Fclipboard-checked-black.webp&w=64&q=75 64w, /_next/image?url=%2Fclipboard-checked-black.webp&w=96&q=75 96w, /_next/image?url=%2Fclipboard-checked-black.webp&w=128&q=75 128w, /_next/image?url=%2Fclipboard-checked-black.webp&w=256&q=75 256w, /_next/image?url=%2Fclipboard-checked-black.webp&w=384&q=75 384w, /_next/image?url=%2Fclipboard-checked-black.webp&w=640&q=75 640w, /_next/image?url=%2Fclipboard-checked-black.webp&w=750&q=75 750w, /_next/image?url=%2Fclipboard-checked-black.webp&w=828&q=75 828w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1080&q=75 1080w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1200&q=75 1200w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1920&q=75 1920w, /_next/image?url=%2Fclipboard-checked-black.webp&w=2048&q=75 2048w, /_next/image?url=%2Fclipboard-checked-black.webp&w=3840&q=75 3840w"
                      style="color: transparent;"
                      width="20"
                    />
                  </div>
                </td>
              </tr>
              <tr
                class="[&>td]:pt-10 [&>td]:first:pt-0"
              >
                <td
                  class="flex flex-col justify-start break-keep pr-2.5 align-top"
                >
                  <h2
                    class="my-0 max-w-lg font-bold"
                  >
                    The Adventures of Huckleberry Finn
                  </h2>
                  <div
                    class="flex flex-col"
                  >
                    <div
                      class="text-sm"
                    >
                      Mark Twain
                    </div>
                    <div
                      class="order-last text-sm max-smaller:order-first"
                      title="Date Completed"
                    >
                      09/02/2025
                    </div>
                  </div>
                </td>
                <td
                  class="align-top text-sm max-smaller:inline"
                >
                  Fiction
                </td>
                <td
                  class="align-top"
                >
                  <div
                    class="mx-auto h-5 w-5"
                  >
                    <img
                      alt="Completed Book Icon"
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      sizes="20px"
                      src="/_next/image?url=%2Fclipboard-checked-black.webp&w=3840&q=75"
                      srcset="/_next/image?url=%2Fclipboard-checked-black.webp&w=16&q=75 16w, /_next/image?url=%2Fclipboard-checked-black.webp&w=32&q=75 32w, /_next/image?url=%2Fclipboard-checked-black.webp&w=48&q=75 48w, /_next/image?url=%2Fclipboard-checked-black.webp&w=64&q=75 64w, /_next/image?url=%2Fclipboard-checked-black.webp&w=96&q=75 96w, /_next/image?url=%2Fclipboard-checked-black.webp&w=128&q=75 128w, /_next/image?url=%2Fclipboard-checked-black.webp&w=256&q=75 256w, /_next/image?url=%2Fclipboard-checked-black.webp&w=384&q=75 384w, /_next/image?url=%2Fclipboard-checked-black.webp&w=640&q=75 640w, /_next/image?url=%2Fclipboard-checked-black.webp&w=750&q=75 750w, /_next/image?url=%2Fclipboard-checked-black.webp&w=828&q=75 828w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1080&q=75 1080w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1200&q=75 1200w, /_next/image?url=%2Fclipboard-checked-black.webp&w=1920&q=75 1920w, /_next/image?url=%2Fclipboard-checked-black.webp&w=2048&q=75 2048w, /_next/image?url=%2Fclipboard-checked-black.webp&w=3840&q=75 3840w"
                      style="color: transparent;"
                      width="20"
                    />
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <button
          aria-label="Show More Books"
          class="mx-auto mt-8 block w-fit rounded-md border px-2 py-1 hover:opacity-50"
        >
          Show More
        </button>
      </div>
    </section>
  </div>
</div>
`;
